File: benchmarks/microbenchmarks/EmailTargets.java
/**
 * Get a list of all users to which a particular user has sent emails.
 *
 * Starting with an email address (the sender), return a list of
 * unique email addresses to which the sender has sent at least one
 * email.
 */

import pmgd.*;
import java.util.Set;
import java.util.HashSet;
import java.lang.String;
import java.lang.management.*;

public class EmailTargets {

    /**
     * Program name.
     */
    static String progName = "EmailTargets";

    /*
     * Tags and properties. 
     */
    static String person_str = "Person";
    static String email_str = "Email";
    static String to_str = "To";
    static String from_str = "From";

    /**
     * The application's main function
     * @param argc the number of command line arguments
     * @param argv the array of command line arguments
     * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
     */
    public static void main(String[] argv)
    {
        boolean timing = false;    ///< Enable timing measurement
        boolean verbose = false;
        int argc = argv.length;

        String graph_name;       ///< Input graph
        String sender_email;     ///< A sender's email address

        int num_recipients;     ///< The number of unique recipients

        // Parse the command line arguments
        int argi = 0;
        while (argi < argc && argv[argi].charAt(0) == '-') {
            if (argv[argi].compareTo("-h") == 0) {
                printUsage();
                return;
            }
            else if (argv[argi].compareTo("-t") == 0)
                timing = true;
            else if (argv[argi].compareTo("-v") == 0)
                verbose = true;
            else {
                System.out.println(progName + ": " + argv[argi] + ": Unrecognized option");
                return;
            }
            argi += 1;
        }
        if (argi + 1 > argc) {
            System.out.println(progName + ": No input graph");
            printUsage();
            return;
        }
        graph_name = argv[argi];
        argi++;
        if (argi + 1 > argc) {
            System.out.println(": No sender's email address");
            printUsage();
            return;
        }
        sender_email = argv[argi];

        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
        long userTime, cpuTime;
        try {
            // Open the graph.
            Graph db = new Graph(graph_name, Graph.OpenOptions.ReadOnly);

            // Start a transaction.
            Transaction tx = new Transaction(db, false, true);

            // Time measurement.
            long startTime = bean.getCurrentThreadUserTime();
            long startTotal = bean.getCurrentThreadCpuTime();

            // Initialize the number of recipients.
            num_recipients = 0;

            // Set of recipients.
            Set<Node> recipients = new HashSet<Node>();

            try {
                // Convert C string to StringIDs.
                StringID person_sid = new StringID(person_str);
                StringID email_sid = new StringID(email_str);
                StringID to_sid = new StringID(to_str);
                StringID from_sid = new StringID(from_str);

                // Get the node representing the sender (we assume it is unique).
                NodeIterator senderi = db.get_nodes(person_sid,
                        new PropertyPredicate(email_sid,
                            PropertyPredicate.Op.Eq,
                            new Property(sender_email)), false);
                // If the sender exists, look for recipients.
                if (!senderi.done()) {
                    for (NodeIterator m = senderi.get_current().get_neighbors(Node.Direction.Incoming, from_sid, false); !m.done(); m.next()) {
                        for (NodeIterator p = m.get_current().get_neighbors(Node.Direction.Outgoing, to_sid, false); !p.done(); p.next()) {
                            Node n = p.get_current();
                            if (!recipients.contains(n)) {
                                recipients.add(n);
                                num_recipients++;
                            }
                        }
                    }
                }
            }
            catch (pmgd.Exception e) {
                e.print();
                return;
            }

            // Time measurement.
            long endTime = bean.getCurrentThreadUserTime();
            long endTotal = bean.getCurrentThreadCpuTime();
            userTime = (endTime - startTime)/1000;
            cpuTime = (endTotal - startTotal)/1000;

            // Print a summary.
            System.out.println(num_recipients);

            // Be more detailed if requested.
            if (verbose)
                for (Node i:recipients)
                    System.out.println(i.get_property(new StringID("Email")).string_value());

            if (timing) {
                long sysTime = cpuTime - userTime;
                float upercent = 100 * (userTime / (float)cpuTime);
                float spercent = 100 * (sysTime / (float)cpuTime);
                System.out.println("Elapsed time is " + cpuTime + " microseconds");
                System.out.println("User time is " + userTime + " microseconds (" + upercent + "%)");
                System.out.println("System time is " + sysTime + " microseconds (" + spercent + "%)");
            }
        }
        catch (pmgd.Exception e) {
            e.print();
            return;
        }

        return;
    }

    /**
     * Print information about how to invoke this program
     */
    static void printUsage()
    {
        System.out.println("Usage: " + progName + " [OPTION]... GRAPH SENDER");
        System.out.println("");
        System.out.println("From GRAPH get a list of all unique email addresses to which");
        System.out.println("SENDER (specified as an email address) has sent emails.");
        System.out.println("");
        System.out.println("  -h  print this help and exit");
        System.out.println("  -t  time the execution");
        System.out.println("  -v  be verbose and print the recipient's email addresses");
    }
}

File: benchmarks/microbenchmarks/emailtargets.cc
/**
 * Get a list of all users to which a particular user has sent emails.
 *
 * Starting with an email address (the sender), return a list of
 * unique email addresses to which the sender has sent at least one
 * email.
 */

#include <stdio.h>
#include <string.h>
#include <iostream>
#include <set>
#include "pmgd.h"
#include "util.h"
#include "neighbor.h"
#include "time.h"

using namespace std;
using namespace PMGD;

/**
 * Program name.
 */
static const char prog_name[] = "emailtargets";

/*
 * Tags and properties. 
 */
const char *person_str = "Person";
const char *email_str = "Email";
const char *to_str = "To";
const char *from_str = "From";

/**
 * Print information about how to invoke this program
 * @param o the put-to stream
 */
void print_usage(ostream &o)
{
    o << "Usage: " << prog_name << " [OPTION]... GRAPH SENDER\n";
    o << "\n";
    o << "From GRAPH get a list of all unique email addresses to which\n";
    o << "SENDER (specified as an email address) has sent emails.\n";
    o << "\n";
    o << "  -h  print this help and exit\n";
    o << "  -t  print elapsed time and usage\n";
    o << "  -v  be verbose and print the recipient's email addresses\n";
}

/**
 * The application's main function
 * @param argc the number of command line arguments
 * @param argv the array of command line arguments
 * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
 */
int main(int argc, char *argv[])
{
    bool timing = false;    ///< Enable timing measurement
    struct timeval t1, t2;
    struct rusage u1, u2;

    bool verbose = false;

    char *graph_name;       ///< Input graph
    char *sender_email;     ///< A sender's email address

    int num_recipients;     ///< The number of unique recipients

    // Parse the command line arguments
    int argi = 1;
    while (argi < argc && argv[argi][0] == '-') {
        if (strcmp(argv[argi], "-h") == 0) {
            print_usage(cout);
            return 0;
        }
        else if (strcmp(argv[argi], "-t") == 0)
            timing = true;
        else if (strcmp(argv[argi], "-v") == 0)
            verbose = true;
        else {
            fprintf(stderr, "%s: %s: Unrecognized option\n", prog_name, argv[argi]);
            return 1;
        }
        argi += 1;
    }
    if (argi + 1 > argc) {
        cerr << prog_name << ": No input graph\n";
        print_usage(cerr);
        return 1;
    }
    graph_name = argv[argi];
    argi++;
    if (argi + 1 > argc) {
        cerr << prog_name << ": No sender's email address\n";
        print_usage(cerr);
        return 1;
    }
    sender_email = argv[argi];

    try {
        // Open the graph.
        Graph db(graph_name, Graph::ReadOnly);

        // Start a transaction.
        Transaction tx(db);

        // Time and usage measurements.
        (void)getrusage(RUSAGE_SELF, &u1);
        (void)gettimeofday(&t1, NULL);

        // Initialize the number of recipients.
        num_recipients = 0;

        // Set of recipients.
        set<Node *> recipients;

        try {
            // Convert C string to StringIDs.
            StringID person_sid(person_str);
            StringID email_sid(email_str);
            StringID to_sid(to_str);
            StringID from_sid(from_str);

            // Get the node representing the sender (we assume it is unique).
            NodeIterator senderi = db.get_nodes(person_sid,
                                                PropertyPredicate(email_sid,
                                                                  PropertyPredicate::Eq,
                                                                  sender_email));
            // If the sender exists, look for recipients.
            if (senderi)
                for (NodeIterator m = get_neighbors(*senderi, Incoming, from_sid, false); m; m.next())
                    for (NodeIterator p = get_neighbors(*m, Outgoing, to_sid, false); p; p.next())
                        if (recipients.find(&*p) == recipients.end()) {
                            recipients.insert(&*p);
                            num_recipients++;
                        }

        }
        catch (Exception e) {
            if (e.num != ReadOnly)
                throw;
        }

        // Time and usage measurements.
        (void)gettimeofday(&t2, NULL);
        (void)getrusage(RUSAGE_SELF, &u2);

        // Print a summary.
        cout << num_recipients << "\n";

        // Be more detailed if requested.
        if (verbose)
            for (auto i:recipients)
                cout << i->get_property("Email").string_value() << "\n";
    }
    catch (Exception e) {
        print_exception(e, stderr);
        return 2;
    }

    if (timing) print_delta(cout, t1, t2, u1, u2);

    return 0;
}
